### PR TITLE
[Issue 743] Sprint burndown by points

### LINKED ISSUE #743: [Task]: Analytics - Add support for sprint burndown by points
### Summary

Currently, sprint burndown is calculated by number of tickets. This would add an option to calculate sprint burndown by points as well.

### Acceptance criteria

- [ ] Sprint burndown can be calculated based on number of points instead of number of tickets

### PR DESCRIPTION
**Pull request by widal001**
## Summary

Adds support for calculating sprint burndown by points instead of tickets, and fixes a few other requested changes that @lucasmbrown-usds made in #685 

Fixes #743

### Time to review: __10 mins__

- Modifies `SprintBurndown` to accept a `unit` parameter which uses the same `Unit` enum as `DeliverablePercentComplete`
- Updates `SprintBurndown.calculate()` to use the `unit` parameter when calculating burndown, allowing us to calculate burndown by sprint and tickets
- Extracts the max number of records returned by the GitHub etl functions and increases it to 10,000 while adding a note to switch to an incremental export and load pattern
- Adds the units to the chart title, axis, and slack post for `SprintBurndown`
- Refactors our use of `Unit` in `DeliverablePercentComplete` to avoid using the hardcoding `"tasks"` or `"points"` as a string outside of the enum
- Updates the README to provide guidance on setting the slack secrets

1. Checkout the PR: `git fetch origin && git checkout issue-743-sprint-burndown-by-points`
2. Follow the "Getting started" instructions if you don't have it set up locally
3. Run: `poetry run analytics calculate sprint_burndown --sprint-file data/sprint-data.json --issue-file data/issue-data.json --sprint "Sprint 10" --unit points --show-results`

### Open questions

- [ ] Do we want to change the unit `tasks` to `tickets`? 

Screenshot 2023-11-30 at 9 44 01 AM
Screenshot 2023-11-30 at 9 43 40 AM

### CODE REVIEW

File: analytics/src/analytics/metrics/percent_complete.py [hunk 1]
```diff
@@ -50,8 +41,8 @@ def calculate(self) -> pd.DataFrame:
         5. Divide closed count by total count to get percent complete
         """
         # get total and closed counts per deliverable
-        df_total = self._get_count_by_deliverable(status="all", unit=self.unit)
-        df_closed = self._get_count_by_deliverable(status="closed", unit=self.unit)
+        df_total = self._get_count_by_deliverable(status="all")
```
Comment (lucasmbrown-usds) [inline analytics/src/analytics/metrics/percent_complete.py:44]: nice to see this cleaned up with the new home for `unit` on `self`

File: analytics/src/analytics/metrics/percent_complete.py [hunk 2]
```diff
@@ -110,23 +102,24 @@ def _get_count_by_deliverable(
...
-        df_agg = df.groupby(self.deliverable_col, as_index=False).agg({unit: "sum"})
-        return df_agg.rename(columns={unit: status})
+        df_agg = df.groupby(self.deliverable_col, as_index=False).agg({unit_col: "sum"})
+        return df_agg.rename(columns={unit_col: status})
 
     def _prepare_result_dataframe_for_plotly(self) -> pd.DataFrame:
         """Stack the open and closed counts self.results for plotly charts."""
         # unpivot open and closed counts so that each deliverable has both
         # an open and a closed row with just one column for count
+        unit = self.unit.value
```
Comment (lucasmbrown-usds) [inline analytics/src/analytics/metrics/percent_complete.py:112]: tiny nitpick: personally i find it a little confusing to name this `unit` when we already have, in other places (i.e. in `self.unit`), a variable called `unit` where the corresponding value that's equal to this is `unit.value`. So in one place, `unit` has a property `value`, and that `value` is called `unit` in another place. 

Consider renaming this version of `unit` to something like `unit_value`?

File: analytics/tests/metrics/test_burndown.py [hunk 3]
```diff
@@ -129,3 +156,46 @@ def test_calculate_burndown_for_current_sprint(self):
             result_row(day=day_3, opened=0, closed=0, delta=0, total=1),
         ]
         assert df.to_dict("records") == expected
+
+
+class TestSprintBurndownByPoints:
```
Comment (lucasmbrown-usds) [inline analytics/tests/metrics/test_burndown.py:161]: amazing tests!!

**Review by lucasmbrown-usds — APPROVED** [review]
Overall looks excellent. On the question of referring to things as either "tasks" or "stories" or something else, I wonder if it makes sense to use the same language as GitHub for consistency and clarity, and call them "issues"? 

But I don't have too strong opinions about this, either way is fine for me. 

Nice work!

File: analytics/src/analytics/metrics/percent_complete.py (hunk 1 above)
Comment (widal001) [reply analytics/src/analytics/metrics/percent_complete.py:44]: Yeah that was a happy side-effect of the switch to using the enum everywhere!

File: analytics/src/analytics/cli.py [hunk 4]
```diff
@@ -66,6 +66,7 @@ def calculate_sprint_burndown(
     issue_file: Annotated[str, ISSUE_FILE_ARG],
```
[the remark this answers is not in the document]
Comment (widal001) [reply analytics/src/analytics/cli.py:66]: Yeah I hadn't seen that before either! But it's the way that [Typer provides hints for options and arguments](https://typer.tiangolo.com/tutorial/arguments/help/#cli-argument-help-panels) in the CLI.

File: analytics/README.md [hunk 5]
```diff
@@ -75,20 +104,21 @@ poetry run analytics export gh_project_data --owner HHS --project 13 --output-fi
...
 
 ```bash
-poetry run analytics calculate sprint_burndown --sprint-file data/sprint-data.json --issue-file data/issue-data.json --sprint @current --show-results
+poetry run analytics calculate sprint_burndown --sprint-file data/sprint-data.json --issue-file data/issue-data.json --sprint @current --unit points --show-results
 ```
 
 A couple of important notes about this command:
 
 - `--sprint @current` In order to calculate burndown, you'll need to specify either `@current` for the current sprint or the name of another sprint, e.g. `"Sprint 10"`
+- `--unit points` In order to calculate burndown based on story points, you pass `points` to the `--unit` option. The other option for unit is `tasks`
```
[the remark this answers is not in the document]
Comment (widal001) [reply analytics/README.md]: Oh I like that! Plus `issues` has the convenient side effect of being the same number of chars as `points`. I replaced references to `tasks` with `issues` in this commit: [refactor: Renames Unit.tasks to Unit.issues](https://github.com/HHS/simpler-grants-gov/pull/782/commits/3d5d56256a118fa0328fb71272347b81f70b0292)

**Pull request merged**